### ISSUE TITLE
Bag.zip

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#1023]
It would be convenient to be able to zip two bags of equal length and partition-lengths together.

This would be particularly handy when splitting and joining bags from an original bag

```python
b = db.from_filenames(...)...
x = b.pluck('x').map(...)
y = b.pluck('y').map(...)
z = b.pluck('z').map(...)
out = db.zip(x, y, z)
```

We might want to be careful to check that the lengths of the interior partitions are the same.

https://groups.google.com/a/continuum.io/d/topic/blaze-dev/NkeyFrOtcWo/discussion

**Comment by mrocklin** [comment]
In simplest case this requires calling `zip` on each of the partitions of each of the input bags together.

In a pretty form it might look something like the following:

```python
def zip(*inputs):
    do various checks
    name = 'zip-' + tokenize(...)
    dsk = {(name, i):  (zip, [(bag.name, i), for bag in inputs]) 
           for i in range(npartitions)}
```

But to be careful about py26 and about laziness we might expand that slightly to the following:

```python
dsk = dict(((name, i), (list, (zip, [(bag.name, i), for bag in inputs]))) 
            for i in range(npartitions))
```

Should we also think about calling `toolz.merge` on the resultant tuple or should that be a separate `map` call?

**Comment by jcorbin** [comment]
Awesome, I have a working local copy of bag zip and will put out a PR for it Soon ™ :-)

**Comment by mrocklin** [comment]
That was fast.  Great!

**Comment by jcorbin** [comment]
Okay recording my working function for now, will get this into a PR as at least a utility in `dask/bag/core.py` even if I can't come up with a good way to hang it off a Bag instance method:
```python
from dask.base import tokenize
from dask.bag.core import reify
from toolz import merge
from toolz.compatibility import zip as the_zip

def bag_zip(*bags):
    npartitions = bags[0].npartitions
    bags_dsk = merge(*(bag.dask for bag in bags))
    assert all(bag.npartitions == npartitions for bag in bags)
    # TODO: do more checks
    name = 'zip-' + tokenize(*bags)
    zip_shard = lambda i: (the_zip,) + tuple((bag.name, i) for bag in bags)
    dsk = dict(((name, i), zip_shard(i)) # do we need to (reify, zip_shard(i)) ?
               for i in range(npartitions))
    return dask.bag.Bag(merge(bags_dsk, dsk), name, npartitions)
```

**Comment by mrocklin** [comment]
In general this looks correct to me.  Adding the call to `reify` is nice in Python 3, where `zip` is lazy.

I think you can probably inline the `zip_shard` function to just be 

```python
dsk = dict(((name, i), (the_zip,) + tuple((bag.name, i) for bag in bags) for i in range(npartitions))
```

And then it would also be good to add some simple tests.

Because it can take several inputs and not just two, I'm personally for keeping this as just a straight function rather than an instance method.

**Comment by jcrist** [comment]
Closed by #1025.
